README.md
<h1 align="center">Welcome to kitchen-slack-bot 👋</h1>
<p>
  <img alt="Version" src="https://img.shields.io/badge/version-1.0.0-blue.svg" />
  <a href="LICENSE.md" target="_blank">
    <img alt="License: MIT" src="https://img.shields.io/badge/License-MIT-blue.svg" />
  </a>
</p>

> A slack bot to provision kitchen's infrastructure used at Nmbrs 🇳.

## Running the app

```bash
# development
$ npm run start

# watch mode
$ npm run start:dev

# run the code as in production mode
$ npm run start:prod
```
## Setup

// TODO

## Contributing

We've set up a separate document for our [contribution guidelines](CONTRIBUTING.md).

## Release History

Please see [CHANGELOG.md](CHANGELOG.md).

## Author(s)

👤 **Nmbrs - Infra Squad**

## 📝 License

This project is [MIT](LICENSE.md) licensed.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Added `CHANGELOG.md` file.

## [v1.0.0] - 2021-07-15

### Added

- First release.
- Added `README.md` file.
- Added github pull request template.

[unreleased]: https://github.com/Nmbrs/template-repo
[v1.0.0]: https://github.com/Nmbrs/template-repo/releases/tag/v1.0.0


src/bot-events.ts
import { App } from '@slack/bolt';
import { createKitchenMessage, destroyKitchenMessage, latestRunMessage as latestRunMessage, unpoliteMessage } from './bot-messages';
import { getRunnerURlRequest, kitchenActionRequest } from './http-requests/github-api';
import config from 'config';
import { firstValueFrom, throwError, timer } from 'rxjs';
import { catchError, exhaustMap, filter, first, take } from 'rxjs/operators';

/**
 * Methods that listen for all slack events that should be interpreted by the
 * slack bot
 *
 * @param app bolt application instance
 *
 */
export async function eventListeners(app: App) {
  const provisionUrl = config.get('github.provisionWorkflowUrl');
  const destroyUrl = config.get('github.destroyWorkflowUrl');

  const provisionPayload = { ref: 'master', inputs: { action: 'create' } };
  const destroyPayload = { ref: 'master', inputs: { action: 'destroy' } };

  app.command('/kitchen', async ({ command, ack, say }) => {
    const date = new Date().toISOString();
    try {
      await ack();

      if (command.text.toLowerCase() === 'status') {
        const result = await getCurrentRunnerData();
        await say(`${latestRunMessage}${result}`);
      } else if (command.text.toLowerCase().includes('mobile create') || command.text.toLowerCase().includes('create mobile')) {
        if (!command.text.toLowerCase().includes('please')) {
          await say(unpoliteMessage);
        } else {
          await say(createKitchenMessage);
          await kitchenActionRequest(provisionUrl, provisionPayload);
          const result = await getCurrentRunnerData(date);
          await say(`${latestRunMessage}${result}`);
        }
      } else if (command.text.toLowerCase().includes('mobile destroy') || command.text.toLowerCase().includes('destroy mobile')) {
        if (!command.text.includes('please')) {
          await say(unpoliteMessage);
        } else {
          await say(destroyKitchenMessage);
          await kitchenActionRequest(destroyUrl, destroyPayload);
          const result = await getCurrentRunnerData(date);
          await say(`${latestRunMessage}${result}`);
        }
      }
    } catch (error) {
      console.error(error.message);
      throw new Error('new error');
    }
  });
}

/**
 * Function that polls the current github runner for a specific workflow
 *
 * @param date specified data fro a given runner
 *
 */
function getCurrentRunnerData(date?: string): Promise<string> {
  const pollingDelay = 2000;
  const pollingInterval = 2000;
  const pollingAttempts = 10;
  pollingDelay.toString();
  return firstValueFrom(
    timer(pollingDelay, pollingInterval).pipe(
      // Tries to read the number of attempts determined by 'pollingAttempts'
      take(pollingAttempts),
      //retrieve the runner data
      exhaustMap(() => {
        return getRunnerURlRequest(date);
      }),
      //return only truth values
      filter((result) => result),
      //Take the first emitted value
      first(),
      //Error Handling
      catchError((error) => {
        return throwError(() => new Error());
      })
    )
  );
}


src/bot-messages.ts
export const customMessage = {
  blocks: [
    {
      type: 'section',
      text: {
        type: 'mrkdwn',
        text: 'Hey! It seems that you need my assistance with your kitchen. What would you like me to do?',
      },
    },
    {
      type: 'actions',
      block_id: 'action_block_kitchen',
      elements: [
        {
          type: 'button',
          text: {
            type: 'plain_text',
            text: 'Provision a new kitchen',
          },
          style: 'primary',
          action_id: 'create_kitchen',
          value: 'create',
        },
        {
          type: 'button',
          text: {
            type: 'plain_text',
            text: 'Destroy my kitchen',
          },
          style: 'danger',
          action_id: 'destroy_kitchen',
          value: 'destroy',
        },
      ],
    },
  ],
};

export const createKitchenMessage = `I cook 🍲, I create, I'm incredibly excited by what I do, I’ve still got a lot to achieve.`;

export const destroyKitchenMessage = `I'll burn your kitchen to the ground 🔥.`;

export const unpoliteMessage = "Where're the manners Johny?";

export const latestRunMessage = "Here's the link for the latest run: "

src/http-requests/github-api.ts
import http from 'http';
import { getConfig } from './../configurations/config';
import axios from 'axios';

export async function kitchenActionRequest(url, data) {
  const token = getConfig<string>('github.personalAccessToken');

  const requestConfiguration = {
    validateStatus: null,
    headers: {
      accept: 'application/vnd.github.v3+json',
      authorization: `Bearer ${token}`,
    },
  };

  let response;
  try {
    response = await axios.post(url, data, requestConfiguration);
  } catch (error) {
    // Network error handling
    console.log(`Error sending HTTP request at ${url}. Error message: ${error.message}.`);
    throw new Error('Internal Error');
  }

  switch (response.status) {
    case 204:
      return response.data;
    case 401:
      console.log(`Invalid credentials sent to the url: ${url}. It could be either invalid credentials or an error.`);
      throw new Error('Internal Error');
    default:
      console.log(`The url: ${url} returned an unexpected response status: "${response.status} - ${http.STATUS_CODES[response.status]}".`);
      throw new Error('Internal Error');
  }
}

export async function getRunnerURlRequest(date?: string) {
  let url = `${getConfig<string>('github.workflowRunnerUrl')}?branch=master&created==>${date}`;
  if (!date) {
    url = `${getConfig<string>('github.workflowRunnerUrl')}?branch=master`;
  }

  const token = getConfig<string>('github.personalAccessToken');

  const requestConfiguration = {
    validateStatus: null,
    headers: {
      accept: 'application/vnd.github.v3+json',
      authorization: `Bearer ${token}`,
    },
  };

  let response;
  try {
    response = await axios.get(url, requestConfiguration);
  } catch (error) {
    // Network error handling
    console.log(`Error sending HTTP request at ${url}. Error message: ${error.message}.`);
    throw new Error('Internal Error');
  }

  switch (response.status) {
    case 200:
      return response.data?.workflow_runs?.[0]?.html_url;
    case 401:
      console.log(`Invalid credentials sent to the url: ${url}. It could be either invalid credentials or an error.`);
      throw new Error('Internal Error');

    default:
      console.log(`The url: ${url} returned an unexpected response status: "${response.status} - ${http.STATUS_CODES[response.status]}".`);
      throw new Error('Internal Error');
  }
}


app.js
const { App } = require('@slack/bolt');
const config = require('config');

// If no environment is defined, sets environment to "development"
if (!process.env.NODE_ENV) {
  process.env.NODE_ENV = 'local';
  console.log(`Warning : Starting the application in "${process.env.NODE_ENV}" mode for debugging purposes.`);
}

const app = new App({
  signingSecret: config.get('slack.signingSecret'),
  token: config.get('slack.botOAuthToken'),
});

app.command('/kitchen', async ({ command, ack, say }) => {
  console.log(command);
  try {
    await ack();
    say(message());
  } catch (error) {
    console.error(error);
  }
});

app.message('', async ({ command, say }) => {
  try {
    say(message());
  } catch (error) {
    console.error(error);
  }
});

// subscribe to 'app_mention' event in your App config
// need app_mentions:read and chat:write scopes
app.event('app_mention', async ({ event, context, client, say }) => {
  try {
    await say(message());
  } catch (error) {
    console.error(error);
  }
});

app.action('create_kitchen', async ({ ack, say }) => {
  const url = config.get('github.provisionWorkflowUrl');
  try {
    await ack();
    await say(`Ok, I'll try my best to create it. 👍`);
    await say(`⚠️ Please, keep in mind that if the kitchen is already in place I can't create it.`);
    const data = {
      ref: 'main',
      inputs: {
        action: 'create',
        environment: 'development',
      },
    };
    kitchenActionRequest(url, data);
  } catch (error) {
    await say('❌ Uh Oh, something went wrong during the process');
  }
});

app.action('destroy_kitchen', async ({ ack, say }) => {
  const url = config.get('github.provisionWorkflowUrl');
  try {
    await ack();
    await say(`Ok, I'll destroy it. 👍`);
    const data = {
      ref: 'main',
      inputs: {
        action: 'destroy',
        environment: 'development',
      },
    };
    kitchenActionRequest(url, data);
  } catch (error) {
    await say('❌ Uh Oh, something went wrong during the process');
  }
});

(async () => {
  const port = config.get('server.port');
  // Start your app
  await app.start(port);
  console.log(`Slack Bot app is listening at http://localhost:${port}`);
})();

async function kitchenActionRequest(url, data) {
  const http = require('http');
  const config = require('config');
  const axios = require('axios');
  const token = config.get('github.personalAccessToken');

  const requestConfiguration = {
    validateStatus: null,
    headers: {
      accept: 'application/vnd.github.v3+json',
      authorization: `Bearer ${token}`,
    },
  };

  let response;
  try {
    response = await axios.post(url, data, requestConfiguration);
  } catch (error) {
    // Network error handling
    console.log(`Error sending HTTP request at ${url}. Error message: ${error.message}.`);
    throw new Error('Internal Error');
  }

  switch (response.status) {
    case 204:
      return response.data;
    case 401:
      basicLog(LogLevels.error, SERVICEID, `Invalid credentials sent to the url: ${url}. It could be either invalid credentials or an error.`);
      throw new Error('Internal Error');
    default:
      console.log(`The url: ${url} returned an unexpected response status: "${response.status} - ${http.STATUS_CODES[response.status]}".`);
      throw new Error('Internal Error');
  }
}

function message() {
  return {
    blocks: [
      {
        type: 'section',
        text: {
          type: 'mrkdwn',
          text: 'Hey! It seems that you need my assistance with your kitchen. What would you like me to do?',
        },
      },
      {
        type: 'actions',
        block_id: 'action_block_kitchen',
        elements: [
          {
            type: 'button',
            text: {
              type: 'plain_text',
              text: 'Provision a new kitchen',
            },
            style: 'primary',
            action_id: 'create_kitchen',
            value: 'create',
          },
          {
            type: 'button',
            text: {
              type: 'plain_text',
              text: 'Destroy my kitchen',
            },
            style: 'danger',
            action_id: 'destroy_kitchen',
            value: 'destroy',
          },
        ],
      },
    ],
  };
}
